### PR TITLE
Carplay for ios 14

### PR DESCRIPTION
**Pull request by tomdye**
Fixes various bugs I found when using this package and adds first class support for ios14. There were a fair few breaking changes from ios13 to ios14 and it was not possible to support both 14 and < 14 within the same package so in my mind it is better to move forward with a version `2.0.0` that supports ios14 and leave the existing version `1.3.1` to support legacy ios versions.

mapTemplateNavigation

- updates readme
- updates the example app
- sets build / deployment target to 14
- improves the on connect / disconnect approach
- adds support for different distance units for travel estimates
- adds support for resizing / tinting manoeuvre symbol images
- remove use of deprecated symbol image
- fixes bug with maps where trip keeps getting recreated
- adds pan events
- adds back button detection to list templates
- adds selected trip preview event
- adds started trip event
- adds leading / trailing navigation bar buttons to map templates
- fixes bar buttons
- fixes map buttons
- fixes travel estimates
- fixes show trip previews
- adds update map buttons functionality
- adds ability to disable bar buttons
- adds trip destination / starting position names
- fixes route choice functionality
- fixes configs which didn't extend `TemplateConfig`
- adds template configs to index barrel file
- fixes pause-reasons enum
- fixes time remaining color
- types many events which were set to `any`

### CODE REVIEW

**Comment by tomdye** [comment]
@birkir please take a look at this and contact me if you'd like to discuss. I've used this to integrate carplay into an existing turn by turn app and got it working pretty well. These are all the bugs / tweaks I made along the way to get it running more stable and with more functionality.
I think moving forward with a v2.0.0 aimed at ios14 is the best approach here.

**Comment by marf** [comment]
> @birkir please take a look at this and contact me if you'd like to discuss. I've used this to integrate carplay into an existing turn by turn app and got it working pretty well. These are all the bugs / tweaks I made along the way to get it running more stable and with more functionality.
> I think moving forward with a v2.0.0 aimed at ios14 is the best approach here.

Hello @tomdye we have a music app similar to Spotify and are trying to integrate CarPlay for it. We are trying to use our fork, the problem is thank when we open the app in our simulator app does not seem to open in CarPlay and if we open it directly in CarPlay it crashes with this message:

`Application does not implement CarPlay template application lifecycle methods in its scene delegate.`

Are we missing something?

Thank you

File: ios/RNCarPlay.m [hunk 1]
```diff
@@ -93,6 +101,41 @@ - (dispatch_queue_t)methodQueue
...
+
+- (UIImage *)imageWithSize:(UIImage *)image convertToSize:(CGSize)size {
+    UIGraphicsBeginImageContext(size);
+    [image drawInRect:CGRectMake(0, 0, size.width, size.height)];
+    UIImage *destImage = UIGraphicsGetImageFromCurrentImageContext();
+    UIGraphicsEndImageContext();
+    return destImage;
+}
+
+- (UIColor *)colorFromHexString:(NSString *)hexString {
```
Comment (birkir) [inline ios/RNCarPlay.m:124]: Could we not use RCTConvert color:UIColor and cast that to rgb?

**Comment by birkir** [comment]
This is awesome.

What was the problem with supporting ios13 with the of statements?

Runtime errors or just didn't build?

File: ios/RCTConvert+RNCarPlay.m [hunk 2]
```diff
@@ -51,16 +51,19 @@ + (CPMapButton*)CPMapButton:(id)json withHandler:(void (^)(CPMapButton * _Nonnul
 
 + (MKMapItem*)MKMapItem:(id)json {
     CLLocationCoordinate2D coordinate = CLLocationCoordinate2DMake([RCTConvert double:json[@"latitude"]], [RCTConvert double:json[@"longitude"]]);
+    NSLog(@">>> Parsing mapItem %@", json);
```
Comment (birkir) [inline ios/RCTConvert+RNCarPlay.m:54]: Maybe have the nslog under a debug flag unless there is a reason for it

**Review by birkir — COMMENTED** [review]
Few feedback items

**Comment by tomdye** [comment]
> This is awesome.
> 
> What was the problem with supporting ios13 with the if statements?
> 
> Runtime errors or just didn't build?

The biggest issue I had was the macro level conditionals in the Objective C. Essentially the PODs / build targets for your package were set to a low iOS level so when they got installed into your react native package they were compiled for the wrong target and thus the statements never got entered. 
I also found that the differing types / rules for the typescript side of the interface were problematic and caused problems.

TLDR: When using the package previously you could not build it to target both pre / post iOS14 so there was no point jumping through hoops and adding complexity to support it. Your app either has to support one of the other.

File: ios/RNCarPlay.m (hunk 1 above)
Comment (tomdye) [reply ios/RNCarPlay.m:124]: My strength does not lie in objective-c land I'm afraid so if you feel there is a better way to do this, please let me know. Happy to try the above out.

File: ios/RCTConvert+RNCarPlay.m (hunk 2 above)
Comment (tomdye) [reply ios/RCTConvert+RNCarPlay.m:54]: I can drop the NSLogs, these are likely left over from my debugging (I kept losing references to the map items due to the trips being recreated every time 😢 )

**Comment by tomdye** [comment]
> > @birkir please take a look at this and contact me if you'd like to discuss. I've used this to integrate carplay into an existing turn by turn app and got it working pretty well. These are all the bugs / tweaks I made along the way to get it running more stable and with more functionality.
> > I think moving forward with a v2.0.0 aimed at ios14 is the best approach here.
> 
> Hello @tomdye we have a music app similar to Spotify and are trying to integrate CarPlay for it. We are trying to use our fork, the problem is thank when we open the app in our simulator app does not seem to open in CarPlay and if we open it directly in CarPlay it crashes with this message:
> 
> `Application does not implement CarPlay template application lifecycle methods in its scene delegate.`
> 
> Are we missing something?
> 
> Thank you

Did you follow the delegate steps in the Readme? https://github.com/birkir/react-native-carplay#installing

**Comment by tomdye** [comment]
@birkir going forward, I think it would be worthwhile converting to `swift` and remove the use of the non-deprecated delegate - though I've not yet looked into the complexities surrounding this.

File: ios/RCTConvert+RNCarPlay.m (hunk 2 above)
Comment (tomdye) [reply ios/RCTConvert+RNCarPlay.m]: @birkir NSLog calls removed

File: ios/RNCarPlay.m (hunk 1 above)
Comment (tomdye) [reply ios/RNCarPlay.m]: @birkir switched over the use `RCTConvert` as you said and made the use of `processColor` optional on the front end, so the user can specify `yellow` / `#ff0000` or `processColor('pink')`. Keeps the API flexible on the TSX side and follows the pattern of resolving image sources for the user etc.

**Comment by tomdye** [comment]
Comments addressed and `prettier` ran against the `example` dir.

**Comment by marf** [comment]
> > > @birkir please take a look at this and contact me if you'd like to discuss. I've used this to integrate carplay into an existing turn by turn app and got it working pretty well. These are all the bugs / tweaks I made along the way to get it running more stable and with more functionality.
> > > I think moving forward with a v2.0.0 aimed at ios14 is the best approach here.
> > 
> > 
> > Hello @tomdye we have a music app similar to Spotify and are trying to integrate CarPlay for it. We are trying to use our fork, the problem is thank when we open the app in our simulator app does not seem to open in CarPlay and if we open it directly in CarPlay it crashes with this message:
> > `Application does not implement CarPlay template application lifecycle methods in its scene delegate.`
> > Are we missing something?
> > Thank you
> 
> Did you follow the delegate steps in the Readme? https://github.com/birkir/react-native-carplay#installing

Yeah I followed them but they are several called, I read somewhere that they are used by Navigation apps only, and if you have Audio app capabilities like in my case they are not called. Maybe is this the problem?

**Comment by tomdye** [comment]
@marf I'm afraid that I have not looked into audio apps yet.

**Comment by kevinlowen1** [comment]
@tomdye I tried using this branch's example with the instructions in the ReadMe of this branch.  It doesn't seem to be working on my machine, think the instructions might need to be updated.

error message is `'React/RCTConvert.h' file not found`

**Comment by tomdye** [comment]
@kevinlowen1 if you already the example etc installed you may need to run `rm -rf node_modules ios/Pods` from the example dir before running the steps. re. POI, I'm afraid I did not look at this template yet as I was working on a navigation app for which POI is not actually a valid template. Once this PR is approved and merged I'm sure contributors like myself can look into the remaining templates currently causing issues.

**Comment by birkir** [comment]
I will release later today, thank you for the effort!

**Comment by tomdye** [comment]
@birkir I'm assuming you don't have a react-native-carplay related discord, but I'm tom#1547 if you wanted to chat at all on discord about this project and future development / releases etc.

**Comment by vvusts** [comment]
@tomdye @birkir sorry I don;t understand. From this PR I see POI template should work on IOS 14 but it crash so is it supported or not?

**Comment by tomdye** [comment]
@vvusts I do not believe I was able to get `POI` to work when I was doing this, but I was writing a turn by turn navigation app when I was making this PR. The `POI` template is only valid for certain types of apps as per the apple developer guide so you may need to check that over and see if your app qualifies.
